File: c++/palindrome_number.cpp
/*
Determine whether an integer is a palindrome. Do this without extra space.
*/

/*
 * A negative number is not a palindrome. O(logN)
*/

#include "helper.h"

class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0) {
            return false;
        }
        int y = 0;
        int tmp = x;
        while (tmp != 0) {
            int residual = tmp % 10;
            tmp /= 10;
            y = y * 10 + residual;
        }
        return x == y;
    }
};

int main() {
    Solution s;
    int x = 0;
    cout << s.isPalindrome(x) << endl;
    x = 12121;
    cout << s.isPalindrome(x) << endl;
    x = 121121;
    cout << s.isPalindrome(x) << endl;
    x = 122321;
    cout << s.isPalindrome(x) << endl;
    return 0;
}

File: python/palindrome_number.py
'''
Determine whether an integer is a palindrome. Do this without extra space.
'''

'''
Assume x is a non-negtive number.
'''
class Solution:
    # @return a boolean
    def isPalindrome(self, x):
        x = str(x)
        first = 0
        last = len(x) - 1
        while first <= last:
            if x[first] != x[last]:
                return False
            first += 1
            last -= 1
        return True

if __name__ == '__main__':
    s = Solution()
    print s.isPalindrome(12321)
    print s.isPalindrome(123321)
    print s.isPalindrome(123)
    print s.isPalindrome(0)
                
